sklearn/utils/validation.py

```diff
@@ -140,7 +140,11 @@ def _num_samples(x):
         if len(x.shape) == 0:
             raise TypeError("Singleton array %r cannot be considered"
                             " a valid collection." % x)
-        return x.shape[0]
+        # Check that shape is returning an integer or default to len
+        if isinstance(x.shape[0], int):
```

Comment:
So it looks like on Python 2.7 Windows 32 bit, we have `isinstance(x.shape[0], int)` is False for sparse arrays. Could it be `np.int` there for some reason?
```py
>>> isinstance(np.int, int)
False
```

Reply:
No actually it works,
```py
>>> isinstance(np.int(42), int)
True
```
but I'm not sure if does for all numpy versions and environements..

Reply:
who cares about legacy python? Or did you want to backport to 0.20.1? Though I guess there is an argument for doing that?

Reply:
I'd recommend `isinstance(x.shape[0], numbers.Integral)`.

Reply:
> who cares about legacy python?

I'm not sure if it's due to python 2, numpy version or the fact that the system is 32 bit: `numbers.Integral` is probably indeed safer..